src/internal/pjs/api_server.go

```diff
@@ -189,3 +176,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 		}
 	}
 }
+
+func (a *apiServer) WalkJob(req *pjs.WalkJobRequest, srv pjs.API_WalkJobServer) (err error) {
+	ctx := pctx.Child(srv.Context(), "walkJob")
+
+	// handle job context and request validation.
+	var id pjsdb.JobID
+	if req.Context != "" {
+		id, err = a.resolveJobCtx(ctx, req.Context)
+		if err != nil {
+			return errors.Wrap(err, "walk job")
+		}
+	} else {
+		// TODO(PJS): do auth.
+	}
+	// TODO(PJS): remove this once auth is implemented.
+	reqID := pjsdb.JobID(req.Job.Id)
+	if id != reqID {
+		log.Error(ctx, "job context token does not match requested job.", zap.Int64("request.Job.ID", req.Job.Id))
+		id = reqID // defaulting to this for testing until auth is implemented.
+	}
+
+	// walk and stream back results.
+	var jobs []pjsdb.Job
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		jobs, err = pjsdb.WalkJob(ctx, sqlTx, id, pjsdb.WalkAlgorithm(req.Algorithm.Number()))
+		return errors.Wrap(err, "with tx")
+	}); err != nil {
+		return errors.Wrap(err, "walk job")
+	}
+	for _, job := range jobs {
+		jobInfo, err := toJobInfo(job)
+		if err != nil {
+			return errors.Wrap(err, "walk job")
+		}
+		resp := &pjs.ListJobResponse{
+			Id:   jobInfo.Job,
+			Info: jobInfo,
+			Details: &pjs.JobInfoDetails{
+				JobInfo: jobInfo,
+			},
+		}
+		if err := srv.Send(resp); err != nil {
+			return errors.Wrap(err, "walk job")
+		}
+	}
+	return nil
+}
+
+func (a *apiServer) resolveJobCtx(ctx context.Context, jobCtx string) (id pjsdb.JobID, err error) {
+	var jobCtxDecoded []byte
+	jobCtxDecoded, err = hex.DecodeString(jobCtx)
+	if err != nil {
+		return 0, errors.Wrap(err, "walk job")
+	}
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		id, err = pjsdb.ResolveJobContext(ctx, sqlTx, jobCtxDecoded)
+		if err != nil {
+			// TODO(PJS): handle NotFound with a special error type that returns the right proto error code.
+			return errors.Wrap(err, "with tx")
+		}
+		return nil
+	}); err != nil {
+		return 0, errors.Wrap(err, "walk job")
+	}
+	return id, nil
+}
+
+func toJobInfo(job pjsdb.Job) (*pjs.JobInfo, error) {
+	jobInfo := &pjs.JobInfo{
+		Job: &pjs.Job{
+			Id: int64(job.ID),
+		},
+		ParentJob: &pjs.Job{
+			Id: int64(job.Parent),
+		},
+		Program: job.Program.HexString(),
+	}
+	for _, filesetID := range job.Inputs {
+		jobInfo.Input = append(jobInfo.Input, filesetID.HexString())
+	}
+	switch {
+	case !job.Queued.IsZero() && job.Processing.IsZero() && job.Done.IsZero():
+		jobInfo.State = pjs.JobState_QUEUED
+	case !job.Queued.IsZero() && !job.Processing.IsZero() && job.Done.IsZero():
+		jobInfo.State = pjs.JobState_PROCESSING
+	case !job.Queued.IsZero() && !job.Processing.IsZero() && !job.Done.IsZero():
+		jobInfo.State = pjs.JobState_DONE
```

Comment:
These cases are kind of hard to follow.  It would be nice to name each one outside the switch statement `queued := !job.Queued.IsZero() && job.Processing.IsZero() && job.Done.IsZero()`, or even a method on job that returns the state enum.

Reply:
I just copied and pasted this code but I think we can just do:
```
case job.Done != time.Time{}:
case job.Processing != time.Time{}:
default
```